de-webapp/src/main/java/org/iplantc/de/server/websocket/ReceiveNotificationsDirect.java

```diff
@@ -28,24 +28,31 @@
 
     private final Logger LOG = LoggerFactory.getLogger(ReceiveNotificationsDirect.class);
 
-    private final Connection connection;
+    private Connection connection;
 
     /**
      * Instantiate new
      */
     public ReceiveNotificationsDirect() {
-       connection = AMQPConnectionManager.getInstance().getConnection();
-       LOG.info("amqp Connection created created!");
+       connection = getConnection();
+       LOG.info("amqp Connection created!");
+    }
+
+    private Connection getConnection() {
+        if (connection == null) {
+            connection = AMQPConnectionManager.getInstance().getConnection();
+        }
+        return connection;
```

Comment:
Does `getConnection` ever return a null pointer then return an actual connection in a later call? I suppose that a null pointer only means that the connection couldn't be established, and not necessarily that the the web socket itself is bad. I wonder if it might be useful to have `ReceiveNotificationsDirect` throw an `IOException` instead and have `MessageHandler` or some higher level object catch the exception and try to reestablish the connection later. Doing that would remove the need to make the `connection` property mutable, but I'm not entirely certain that it wouldn't create some other problems.

Reply:
I was testing this by updating the firewall rules on the AMQP box.  So with that in mind, it was possible for getConnection to start returning an actual connection after having returned a null pointer once I updated the firewall to stop blocking me.
I'm definitely not the most familiar with this part of the code, so if there's a better way to handle this I'm definitely open 😄

Reply:
Ah, that makes sense. I also learned a little bit about how the code works since writing this comment, so it makes more sense to me now. 😄  Does the client-side code attempt to reconnect if the web socket is closed? If it does then one approach might be to close the web socket if the AMQP connection can't be established.

Reply:
Yes, there's a check [here](https://github.com/cyverse-de/ui/blob/master/de-lib/src/main/java/org/iplantc/de/desktop/client/presenter/DesktopPresenterImpl.java#L221-L224) when the websocket gets closed.  If the user wasn't trying to log out, then we try to reopen the websocket and then all this connection logic gets re-run again.

It would probably be cleaner to close the websocket right away in `AMQPConnectionManager` like you mentioned, but then I'm not sure how to get access to the websocket from that stage in order to close it?  Or maybe there's a different way to close it I'm not aware of?

Reply:
Ah, good news. One way to do it would be to throw an exception in `ReceiveNotificationsDirect`'s constructor if the connection can't be established. Then `MessageHandler.onOpen` can catch the exception and close the web socket. Another option would be to have `ReceiveNotificationsDirect.createChannel` simply return null if the `connection` property is null. If you did that then the code that you've added to `MessageHandler.onOpen` would detect the problem and close the web socket.

I think I like the idea of having the constructor throw an `IOException` if the connection can't be established. In that case, the error is being handled closer to its source. Other options would definitely work, though.

Reply:
OK perfect, I'll give that a go tomorrow - thanks for the help!